File: Source/Entities/Crossover/BalloonTripCollectable.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.KoseiHelper.Entities.Crossover;

[CustomEntity("KoseiHelper/BalloonTripCollectable")]
[Tracked]
public class BalloonTripCollectable : Entity
{

    public string counterName;
    public int pointsGiven;
    public string sound;
    public bool outline;

    private Sprite sprite;
    private EntityID id;
    private bool canReappear = true;
    private bool collectionEffects = false;
    public static ParticleType balloonParticle = Player.P_Split;
    public bool multiplicative;
    public BalloonTripCollectable(EntityData data, Vector2 offset, EntityID id)
    {
        Depth = data.Int("depth", -100);
        Position = data.Position + offset;
        this.id = id;
        counterName = data.Attr("counterName", "koseiHelper_balloonPoints");
        pointsGiven = data.Int("pointsGiven", 100);
        outline = data.Bool("outline", false);
        sound = data.Attr("sound", "event:/KoseiHelper/Crossover/BalloonExplode");
        canReappear = data.Bool("canReappear", true);
        collectionEffects = data.Bool("collectionEffects", false);
        multiplicative = data.Bool("multiplicative", false);
        Add(sprite = GFX.SpriteBank.Create(data.Attr("spriteID", "koseiHelper_balloonTripCollectable")));
        sprite.CenterOrigin();
        balloonParticle.Color = balloonParticle.Color2 = KoseiHelperUtils.ParseHexColor(data.Values.TryGetValue("particleColor", out object c1) ? c1.ToString() : null,
            Color.FromNonPremultiplied(231, 0, 91, 255));
        sprite.Play("balloon");
        base.Collider = new Hitbox(data.Float("hitboxWidth",12f), data.Float("hitboxHeight",12f), data.Float("hitboxXOffset",-6f), data.Float("hitboxYOffset",-8f));
        Add(new PlayerCollider(OnPlayer));
    }

    public void OnPlayer(Player player)
    {
        Audio.Play(sound);
        if (multiplicative)
            player.level.Session.SetCounter(counterName, player.level.Session.GetCounter(counterName) * pointsGiven);
        else
            player.level.Session.SetCounter(counterName, player.level.Session.GetCounter(counterName) + pointsGiven);
        if (collectionEffects)
        {
            Input.Rumble(RumbleStrength.Medium, RumbleLength.Medium);
            Celeste.Freeze(0.05f);
            SceneAs<Level>().ParticlesBG.Emit(balloonParticle, 4, Center, Vector2.One * 4f);
        }
        RemoveSelf();
        if (!canReappear)
            SceneAs<Level>().Session.DoNotLoad.Add(id);
    }

    public override void Render()
    {
        if (outline)
            sprite.DrawOutline();
        base.Render();
    }
}

File: Loenn/entities/Crossover/BalloonTripCollectable.lua
local mods = require("mods")
local depths = mods.requireFromPlugin("libraries.depths")

local BalloonTripCollectable = {}

BalloonTripCollectable.name = "KoseiHelper/BalloonTripCollectable"
BalloonTripCollectable.texture ="objects/KoseiHelper/Crossover/BalloonTripCollectable/balloon00"

function BalloonTripCollectable.depth(room,entity)
	return entity.depth
end

BalloonTripCollectable.placements = {
	{
		name = "BalloonTripCollectable",
		data = {
			depth = -100,
			pointsGiven = 100,
			counterName = "koseiHelper_balloonPoints",
			spriteID = "koseiHelper_balloonTripCollectable",
			sound = "event:/KoseiHelper/Crossover/BalloonExplode",
			outline = false,
			canReappear = true,
			collectionEffects = false, 
			multiplicative = false,
			hitboxWidth = 12,
			hitboxHeight = 12,
			hitboxXOffset = -6,
			hitboxYOffset = -8,
			particleColor = "e7005bff",
		}
	}
}

BalloonTripCollectable.fieldInformation = {
	pointsGiven = {
		fieldType = "integer"
	},
	hitboxWidth = {
		fieldType = "integer"
	},
	hitboxHeight = {
		fieldType = "integer"
	},
	hitboxXOffset = {
		fieldType = "integer"
	},
	hitboxYOffset = {
		fieldType = "integer"
	},
	depth = {
        fieldType = "integer",
        options = depths.addDepths(depths.getDepths(), {}),
        editable = true
    },
	particleColor = {
		fieldType = "color",
		useAlpha = true
	}
}

--function BalloonTripCollectable.selection(room, entity)
--    local width, height = 16, 16
--    return utils.rectangle(entity.x - width / 2, entity.y - height / 2, width, height)
--end

return BalloonTripCollectable
